girder_volview/dicom.py

```diff
@@ -0,0 +1,139 @@
+            # A pydicom Sequence is a nested list of Datasets, which is too complicated to flatten
+            # now
+            raise ValueError("Cannot coerce a Sequence")
+        return list(map(_coerceValue, value))
+
+    # pydicom does not treat the PersonName type as a subclass of a text type
+    if isinstance(value, pydicom.valuerep.PersonName):
+        return str(value)
+
+    # Many pydicom value types are subclasses of base types; to ensure the value can be serialized
+    # to MongoDB, cast the value back to its base type
+    for knownBaseType in {
+        datetime.datetime,
+        datetime.date,
+        datetime.time,
+        int,
+        float,
+        str,
+    }:
+        if isinstance(value, knownBaseType):
+            return knownBaseType(value)
+
+    if isinstance(value, bytes):
+        if b"\x00" in value:
+            raise ValueError("Binary data with null")
+        # For binary data, see if it can be coerced further into utf8 data.  If
+        # not, mongo won't store it, so don't accept it here.
+        try:
+            value.decode()
+            return value
+        except UnicodeDecodeError:
+            raise ValueError("Binary data that cannot be stored as utf-8")
+
+    raise ValueError("Unknown type", type(value))
+
+
+def _coerceMetadata(dataset):
+    metadata = {}
+
+    # Use simple iteration instead of "dataset.iterall", to prevent recursing into Sequences, which
+    # are too complicated to flatten now
+    # The dataset iterator is
+    #   for tag in sorted(dataset.keys()):
+    #       yield dataset[tag]
+    # but we want to ignore certain exceptions of delayed data loading, so
+    # we iterate through the dataset ourselves.
+    for tag in dataset.keys():
+        try:
+            dataElement = dataset[tag]
+        except OSError:
+            continue
+        if dataElement.tag.element == 0:
+            # Skip Group Length tags, which are always element 0x0000
+            continue
+
+        # Use "keyword" instead of "name", as the keyword is a simpler and more uniform string
+        # See: http://dicom.nema.org/medical/dicom/current/output/html/part06.html#table_6-1
+        # For unknown / private tags, allow pydicom to create a string representation like
+        # "(0013, 1010)"
+        tagKey = (
+            dataElement.keyword
+            if dataElement.keyword and not dataElement.tag.is_private
+            else str(dataElement.tag)
+        )
+
+        try:
+            tagValue = _coerceValue(dataElement.value)
+        except ValueError:
+            # Omit tags where the value cannot be coerced to JSON-encodable types
+            continue
+
+        if tagValue == "" or tagValue == b"":
+            continue
+        if sys.getsizeof(tagValue) > MAX_TAG_SIZE:
+            continue
+
+        metadata[tagKey] = tagValue
+
+    return metadata
+
+
+def _parseFile(f):
+    try:
+        # download file and try to parse dicom
+        with File().open(f) as fp:
```

Comment:
In a different context I was importing a file that an issue and this open failed (it looks like the file had been deleted between scanning for import and actually handling the event).  This failure was a GirderException.  We might want to catch `GirderException` (besides `InvalidDicomError`) and return None.

Reply:
Timing funny stuff eh?  I added `GirderException` to the except clause.
```
    except (pydicom.errors.InvalidDicomError, GirderException):
```

Reply:
Continuing on this theme: when we add a CLI task, we create a girder item that has a "link" file.  These files don't open via `File().open(file)`, and I get a `'NoneType' object has no attribute 'open'` error (this is an `AttributeError`).

Reply:
The hits keep coming... Doing extension and mime checks before the File().open now like `large_image`
```
def addDicomTagsToItem(file):
    """
    Add DICOM tags to Item metadata.
    """
    knownMime = file.get("mimeType") == MIME_TYPE
    exts = [ext.split()[0] for ext in file.get("exts") if ext]
    knownExt = set(exts[-1:]).intersection(EXTENSIONS)
    if not knownExt and not knownMime:
        return
    itemId = file.get("itemId")
    if not itemId:
        return

    dicomMetadata = _parseFile(file)
```

Reply:
Sadly, only using extensions and mimetypes will miss imported files where we haven't detected the mime type appropriately and are named like `1.2.826.0.1.3680043.10.511.3.12939397909077114107649285416788872` (a dicom UID without any extension).  Worse, we have sample dicoms from medical.nema.org which have names like `4_0` (see the 3DHISTECH samples on ftp://medical.nema.org/MEDICAL/Dicom/DataSets/WG26/WG26Connectathon2023_ECDP/).

But, at least now I'm not seeing any exceptions being thrown and unhandled.

Reply:
Gotchata.  Removed extension and mimeType checks.  Added a `if "linkUrl" in file:` check to skip link files.

Also, 
- Updating the mimeType of a file if the DICOM parse was successful and current file mimeType is `None` or `application/octet-stream`
- Using mimeType when checking for VolView loadable files